keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,194 @@ the remaining requests in that queue start getting faster service.  In
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorithm:
+- as soon as we choose the request to dispatch (i.e. the queue from which
+  the first request should be dispatched), we start accumulating concurrency
+  units until we accumulate `<width>` and only then dispatch the request.
+  In other words, if the chosen request has width `<width>` and there
+  are less then `<width>` available seats, we don't dispatch any other request
+  (at a given priority level) until we will have `<width>` available seats
+  at which point we dispatch this request.
+  Such approach (as opposed to dispatching individual concurrency units
+  independently one-by-one) allows us to not waste too many seats and avoid
+  deadlocks if we would be dispatching seats to multiple LIST requests
+  without having enough of them for a given priority level.
+- however, to ensure fairness (especially over longer period of times)
+  we need to change how virtual time is advanced too.  We will change the
+  semantics of virtual time tracked by the queues to correspond to work,
+  instead of just wall time.  That means when we estimate a request's
+  virtual duration, we will use `estimated width x estimated latency` instead
+  of just estimated latecy.  And when a request finishes, we will update
+  the virtual time for it with `seats x actual latency` (note that seats
+  will always equal the estimated width, since we have no way to figure out
+  if a request used less concurrency than we granted it).
+
+However, now the queueing mechanism also requires adjustment.  So far,
+when putting a request into the queue, we were choosing the shortest queue.
+It worked, because it well proxied the total cost of processing all requests
+in that queue.
+After the above changes, the size of the queue is no longer correctly
+approximating the cost of processing the queued items.  Given that the
+total cost of processing a request is now `<width> x <processing latency>`,
+the weight of the queue should now reflect that.
+As a result, for the purpose of estimating the cost of queue, we are going
```

Comment:
OK I think you can remove this sentence now.

Reply:
Done